Add support for Seeed Studio ESP32C3 Xiao

dpproto opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md.
-->

### Contribution description

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->
Add support for Seeed Studio ESP32C3 Xiao. See issue #21266.

### Testing procedure

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->
Tested with the following examples:
- blinky
- gnrc_networking (Wi-Fi)
- nimble_gatt (Bluetooth)

### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

crasbe commented:
Just yesterday I thought about buying a Xiao-ESP32C3 on Berrybase, but they didn't have the Xiao-ESP32S3 in stock (for which another PR is open) but then decided against it because there is no PR open for it. What a coincidence :D

I'll order one and give this a test soon. But so far this looks good, I'll have to take a deeper look when the hardware is here.

<hr/>

You have some files where the copyright notice is still saying "Gunar Schorcht", but the author in the doxygen block is already updated.

Also please update the commit message with the typical format: https://github.com/RIOT-OS/RIOT/blob/master/CONTRIBUTING.md#commit-conventions
For this commit, you could write "boards/seeedstudio-xiao-esp32c3: add support" for example.

dpproto commented:
> And according to #21257, please change your board documentation from `doc.txt` to `doc.md` (getting rid of the surrounding C-comments).

The problem is that a bookmark named "doc" at the 1st level of the documentation tree (RIOT/doc/doxygen/html/md_boards_2seeedstudio-xiao-esp32c3_2doc.html). You can see that for other boards as well in the online documentation, e.g. https://doc.riot-os.org/md_boards_2nucleo-f413zh_2doc.html

How do I avoid this ?

crasbe commented:
> > And according to #21257, please change your board documentation from `doc.txt` to `doc.md` (getting rid of the surrounding C-comments).
> 
> The problem is that a bookmark named "doc" at the 1st level of the documentation tree (RIOT/doc/doxygen/html/md_boards_2seeedstudio-xiao-esp32c3_2doc.html). You can see that for other boards as well in the online documentation, e.g. https://doc.riot-os.org/md_boards_2nucleo-f413zh_2doc.html
> 
> How do I avoid this ?

This is a doxygen bug: https://github.com/RIOT-OS/RIOT/issues/21220#issuecomment-2700537054
It is already resolved and once RIOT switches to the latest Doxygen version (which is yet to be released), it won't be an issue anymore.

I'm glad that you noticed it though, that means you test the documentation :D 👍

crasbe commented on boards/seeedstudio-xiao-esp32c3/doc.md:
```diff
@@ -0,0 +1,164 @@
+BOARD=seeedstudio-xiao-esp32c3 make ...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Almost all GPIOs are broken out and can be used for different peripherals:
```
I don't think this is true for the Xiao, is it?

crasbe commented on boards/seeedstudio-xiao-esp32c3/include/board.h:
```diff
@@ -0,0 +1,72 @@
+/**
+ * @brief   Default button GPIO mode definition
+ *
+ * Since the GPIO of the button is pulled up with an external resistor, the
+ * mode for the GPIO pin has to be GPIO_IN.
+ */
+#define BTN0_MODE       GPIO_IN_PU
```
The documentation contradicts the actual definition.

dpproto replied:
Could you be more specific?

- From the [schematic](https://files.seeedstudio.com/wiki/XIAO_WiFi/Resources/Seeeduino-XIAO-ESP32C3-SCH.pdf), I can tell that D9 = GPIO9 is connected to the BOOT button and to a pull-up resistor.
- From the [ESP32C3 Tech Ref Manual](https://www.espressif.com/sites/default/files/documentation/esp32-c3_technical_reference_manual_en.pdf) (§7.1 p198), I can tell that:
    - GPIO9 is a strapping pin, connected to an internal pull-up resistor
    - GPIO9 can be used as a user button since _"After the reset is released, the strapping pins work as normal-function pins."_

The `GPIO_IN_PU` mode here might be considered redundant or explanatory, but it's an opinion: repeating the default configuration will not harm. Will it ?

crasbe replied:
Oh, I just meant that the comment says `GPIO_IN` and the define says `GPIO_IN_PU`.
I'm totally fine with enabling the internal pullup, but then the comment should be changed to `GPIO_IN_PU`.

Sorry for not making this clearer.

gschorcht replied:
IMHO the button mode should be `GPIO_IN` if an external pull-up is connected. Otherwise, you would have two pull-up resistors in parallel if `BTN0_MODE` is used to initialize that pin like in `tests/buttons/main.c` or `tests/periph/pm/main.c`. Having the two pull-up resistors in parallel gives a smaller resulting resistor value than the lowest single one. The existence of the external pull-up should be only documented.

crasbe commented on boards/seeedstudio-xiao-esp32c3/include/periph_conf.h:
```diff
@@ -0,0 +1,156 @@
+/**
+ * @name   PWM channel configuration
+ *
+ * For generic boards, two PWM devices are configured. These devices
+ * contain all GPIOs that are not defined as I2C, SPI or UART for this board.
+ * Generally, all outputs pins could be used as PWM channels.
```
This should probably be adapted for the Xiao specifically as well.

crasbe commented on boards/seeedstudio-xiao-esp32c3/include/periph_conf.h:
```diff
@@ -0,0 +1,156 @@
+/**
+ * @name   I2C configuration
+ *
+ * For generic boards, only one I2C interface I2C_DEV(0) is defined.
```
This should probably be adapted for the Xiao specifically as well.

crasbe commented on boards/seeedstudio-xiao-esp32c3/include/periph_conf.h:
```diff
@@ -0,0 +1,156 @@
+/**
+ * @brief   Declaration of GPIOs that can be used as ADC channels
+ *
+ * For generic boards, all ADC pins that have broken out are declared as ADC
+ * channels.
```
This should probably be adapted for the Xiao specifically or removed. For the Xiao, only three pins can be used as ADC channels.

crasbe commented on boards/seeedstudio-xiao-esp32c3/include/periph_conf.h:
```diff
@@ -0,0 +1,156 @@
+#endif
+
+/**
+ * @name    ADC and DAC channel configuration
```
There is no channel definition for the DAC. I'm not familiar with the ESP32s, is a dedicated DAC configuration required or is that done implicitly through the ADC configuration?

gschorcht replied:
ESP32-C3 doesn't have a DAC peripheral.

dpproto replied:
The ESP32C3 doesn't have a DAC.

crasbe reviewed (commented):
My Xiao ESP32C3 should arrive this week, then I can get this tested with hardware.

For the time being I have some small remarks about the documentation, but nothing major.

crasbe commented:
One thing I forgot to mention in the review: @mguetschow introduced the `xiao_shield` feature in #20980. It would be good to have this in this PR as well, but obviously you can't add it until mguetschows PR is merged...

crasbe commented:
> > One thing I forgot to mention in the review: @mguetschow introduced the `xiao_shield` feature in #20980. It would be good to have this in this PR as well, but obviously you can't add it until mguetschows PR is merged...
> 
> I searched the "shield" string in the PR you mention, but got no hit. What is the link with the present PR?

https://github.com/RIOT-OS/RIOT/pull/20980/files#diff-999ca1d04a554e7fb0bfda8adf6edb01fe4a67356449490b16e7c4716bde1c84R16

This is what you'd have to add to your PR, but before it is merged it'll probably create a Make error.

crasbe commented:
So today the board finally arrived and I could give this a test. It appears as if it loses some data occasionally.
But I'm sure that that's not because of your board. Either it's caused by my system with USBipd or it's a bug in the ESP32C3 code.
I'll give it a test tomorrow with my ThinkPad that runs Linux.

<details><summary>`tests/sys/shell`</summary>

```
/home/cbuec/RIOTstuff/riot-esp32c3/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200" -ln "/tmp/pyterm-cbuec" -rn "2025-03-13_21.53.21-tests_shell-seeedstudio-xiao-esp32c3"
2025-03-13 21:53:21,369 # Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
help
2025-03-13 21:53:25,808 # help
2025-03-13 21:53:25,810 # Command              Description
2025-03-13 21:53:25,811 # -------------------------------
2025-03-13 21:53:25,812 # bufsize              Get the shell's buffer size
2025-03-13 21:53:25,813 # start_test           starts a test
2025-03-13 21:53:25,814 # end_test             ends a                 prints the input command
2025-03-13 21:53:25,815 # empty                print nothing on command
2025-03-13 21:53:25,816 # periodic             periodically print command
2025-03-13 21:53:25,816 # xfa_test1            xfa test command 1
2025-03-13 21:53:25,817 # xfa_test2            xfa test command 2
2025-03-13 21:53:25,818 # app_metadata         Returns application metadata
2025-03-13 21:53:25,818 # pm             tion about running threads.
2025-03-13 21:53:25,819 # version              Prints current RIOT_VERSION
2025-03-13 21:53:25,819 # reboot               Reboot the node
>
```

</details>

The ADC test works and the lines go to the values they're supposed to go to.

<details><summary>tests/periph/adc</summary>

```
/home/cbuec/RIOTstuff/riot-esp32c3/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200" -ln "/tmp/pyterm-cbuec" -rn "2025-03-13_21.55.48-tests_adc-seeedstudio-xiao-esp32c3"
2025-03-13 21:55:48,802 # Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
2025-03-13 21:55:49,810 # ADC_LINE(0): 13  51  195  732    -1    -1
2025-03-13 21:55:49,810 # ADC_LINE(1): 13  50  191  716    -1    -1
2025-03-13 21:55:49,811 # ADC_LINE(2): 11  45  171  653   -1    -1
2025-03-13 21:55:49,811 #
2025-03-13 21:55:49,811 # ADC_LINE(0): 13  52  196  744    -1    -1
2025-03-13 21:55:49,811 # ADC_LINE(1): 13  51  191  730    -1    -1
2025-03-13 21:55:49,812 # ADC_LINE(2): 12  45  174  655  ADC_LINE(0): 13  52  196  736    -1    -1
2025-03-13 21:55:49,812 # ADC_LINE(1): 13  50  191  720    -1    -1
2025-03-13 21:55:49,812 # ADC_LINE(2): 11  45  174  664    -1    -1
2025-03-13 21:55:49,813 #
2025-03-13 21:55:49,813 # ADC_LINE(0): 13  52  198  746    -1    -1
2025-03-13 21:55:49,813 # ADC_LINE(1): 13  50  191  727    -1    -1
2025-03-13 21:55:49,814 # ADC_LINE(2): 11  45  174 ADC_LINE(0): 14  52  198  748    -1    -1
2025-03-13 21:55:49,814 # ADC_LINE(1): 13  50  191  728    -1    -1
2025-03-13 21:55:49,814 # ADC_LINE(2): 12  45  174  664    -1    -1
2025-03-13 21:55:49,815 #
2025-03-13 21:55:49,815 # ADC_LINE(0): 13  52  198  744    -1    -1
2025-03-13 21:55:49,816 # ADC_LINE(1): 13  50  193  729    -1    -1
2025-03-13 21:55:49,816 # ADC_LINE(2): 12  45  174  664    -1    -1
2025-03-13 21:55:49,816 #
2025-03-13 21:55:49,817 # ADC_LINE(0): 13  52  198  745    -1    -1
2025-03-13 21:55:49,817 # ADC_LINE(1): 13  50  190  717    -1    -1
2025-03-13 21:55:49,818 # ADC_LINE(2): 11  45  171  661ADC_LINE(0): 13  52  195  735    -1    -1
2025-03-13 21:55:49,818 # ADC_LINE(1): 13  50  190  716    -1    -1
2025-03-13 21:55:49,819 # ADC_LINE(2): 11  45  174  653    -1 ADC_LINE(0): 13  51  195  732    -1    -1
2025-03-13 21:55:49,819 # ADC_LINE(1): 13  49  187  712    -1    -1
2025-03-13 21:55:49,820 # ADC_LINE(2): 11  44  171  ADC_LINE(0): 13  51  194  732    -1    -1
2025-03-13 21:55:49,820 # ADC_LINE(1): 13  50  187  712    -1    -1
2025-03-13 21:55:49,856 # ADC_LINE(2): 11  45  171  ADC_LINE(0): 13  51  194  730    -1    -1
2025-03-13 21:55:49,857 # ADC_LINE(1): 13  49  187  712    -1    -1
2025-03-13 21:55:49,858 # ADC_LINE(2): 11  45  174  662    -1    -1
2025-03-13 21:55:49,858 #
2025-03-13 21:55:49,958 # ADC_LINE(0): 13  51  195  732    -1    -1
2025-03-13 21:55:49,959 # ADC_LINE(1): 13  49  189  712    -1    -1
2025-03-13 21:55:49,959 # ADC_LINE(2): 11  45  174  662    -1    -1
2025-03-13 21:55:49,959 #
2025-03-13 21:55:50,060 # ADC_LINE(0): 13  51  195  734    -1    -1
2025-03-13 21:55:50,062 # ADC_LINE(1): 13  50  190  720    -1    -1
2025-03-13 21:55:50,163 # ADC_LINE(2): 11  45  174  6ADC_LINE(0): 13  52  198  744    -1    -1
2025-03-13 21:55:50,164 # ADC_LINE(1): 13  50  190  716    -1    -1
2025-03-13 21:55:50,263 # ADC_LINE(2): 12  45  174  665    -1ADC_LINE(0): 14  52  198  747    -1    -1
2025-03-13 21:55:50,264 # ADC_LINE(1): 13  50  190  720    -1    -1
2025-03-13 21:55:50,365 # ADC_LINE(2): 12  45  175  666    -1    -ADC_LINE(0): 14  52  199  750    -1    -1
2025-03-13 21:55:50,366 # ADC_LINE(1): 13  50  194  729    -1    -1
2025-03-13 21:55:50,466 # ADC_LINE(2): 12  45  174  ADC_LINE(0): 14  52  198  748    -1    -1
2025-03-13 21:55:50,467 # ADC_LINE(1): 13  51  194  729    -1    -1
2025-03-13 21:55:50,567 # ADC_LINE(2): 12  45  174  665    -1ADC_LINE(0): 14  52  199  748    -1    -1
2025-03-13 21:55:50,568 # ADC_LINE(1): 13  50  191  728    -1    -1
2025-03-13 21:55:50,568 # ADC_LINE(2): 12  45  17   -1
2025-03-13 21:55:50,568 #
2025-03-13 21:55:50,669 # ADC_LINE(0): 13  52  198  744    -1    -1
2025-03-13 21:55:50,669 # ADC_LINE(1): 13  50  190  716    -1    -1
2025-03-13 21:55:50,771 # ADC_LINE(2): 11  45  172  655    -1    ADC_LINE(0): 13  52  195  743    -1    -1
2025-03-13 21:55:50,773 # ADC_LINE(1): 13  50  190  715    -1    -1
2025-03-13 21:55:50,872 # ADC_LINE(2): 11  45  171 ADC_LINE(0): 13  51  195  734    -1    -1
2025-03-13 21:55:50,873 # ADC_LINE(1): 13  49  189  711    -1    -1
2025-03-13 21:55:50,976 # ADC_LINE(2): 11  45  1ADC_LINE(0): 13  51  194  731    -1    -1
2025-03-13 21:55:50,976 # ADC_LINE(1): 13  50  190  715    -1    -1
2025-03-13 21:55:51,075 # ADC_LINE(2): 11  44  171  ADC_LINE(0): 13  51  194  730    -1    -1
2025-03-13 21:55:51,078 # ADC_LINE(1): 13  49  187  712    -1    -1
2025-03-13 21:55:51,079 # ADC_LINE(2): 11  45  171  652    -1    -1
2025-03-13 21:55:51,079 #
2025-03-13 21:55:51,177 # ADC_LINE(0): 13  51  194  732    -1    -1
2025-03-13 21:55:51,178 # ADC_LINE(1): 13  50  190  714    -1    -1
2025-03-13 21:55:51,179 # ADC_LINE(2): 11  45  174  664    -1    -1
2025-03-13 21:55:51,279 # ADC_LINE(0): 13  52  197  735    -1    -1
2025-03-13 21:55:51,280 # ADC_LINE(1): 13  50  190  715    -1    -1
2025-03-13 21:55:51,281 # ADC_LINE(2): 12  45  174  663    -1    -1
2025-03-13 21:55:51,281 #
2025-03-13 21:55:51,380 # ADC_LINE(0): 13  52  198  745    -1    -1
2025-03-13 21:55:51,382 # ADC_LINE(1): 13  50  190  716    -1    -1
2025-03-13 21:55:51,383 # ADC_LINE(2): 12  45  175  666    -1    -1
2025-03-13 21:55:51,383 #
```
</details>

crasbe commented:
Oh, but before merging, please add the `xiao_shield` feature. PR #20980 is merged now.

dpproto commented:
> Oh, but before merging, please add the `xiao_shield` feature. PR #20980 is merged now.

Do I just need to add `FEATURES_PROVIDED += xiao_shield` to Makefile.features ? Do I need to rebase on master first ?

crasbe commented:
> Do I just need to add `FEATURES_PROVIDED += xiao_shield` to Makefile.features ? Do I need to rebase on master first ?

Yes, you have to rebase it but I don't think there would be any conflicts if you changed the `Makefile.features` file first. Make would just complain.

dpproto commented:
> > Do I just need to add `FEATURES_PROVIDED += xiao_shield` to Makefile.features ? Do I need to rebase on master first ?
> 
> Yes, you have to rebase it but I don't think there would be any conflicts if you changed the `Makefile.features` file first. Make would just complain.

I added the feature and rebased on master. Thanks for your help.
